numpy/core/src/umath/dispatching.c

```diff
@@ -0,0 +1,697 @@
+                matches = NPY_FALSE;
+                break;
+            }
+            if (given_dtype == NULL) {
+                /*
+                 * If the (input) was not specified, this is a reduce-like
+                 * operation.  Some ufuncs may have non-trivial promotion
+                 * (e.g. add/multiply ensure high precision).
+                 * Continuing here matches promoters: those can deal with that.
+                 * If we allow this path for ArrayMethod, the person
+                 * registering will have it works sense for the ufunc, a
+                 * counter example is `(BoolLike, Bool, Bool)` for `add`.
+                 * It should resolve to an integer result (sum the bools)
+                 * in a reduction. But the ArrayMethod cannot work with that
+                 * (NumPy will prevent it to ensure correctness).
+                 */
+                continue;
+            }
+            int subclass = PyObject_IsSubclass(
+                    (PyObject *)given_dtype, (PyObject *)resolver_dtype);
+            if (subclass < 0) {
+                return -1;
+            }
+            if (!subclass) {
+                matches = NPY_FALSE;
+                break;
+            }
+            /*
+             * TODO: Could consider allowing reverse subclass relation, i.e.
+             *       the operation DType passed in to be abstract.  That
+             *       definitely is OK for outputs (and potentially useful,
+             *       you could enforce e.g. an inexact result).
+             *       It might also be useful for some stranger promoters.
+             */
+        }
+        if (!matches) {
+            continue;
+        }
+
+        /* The resolver matches, but we have to check if it is better */
+        if (best_dtypes != NULL) {
+            int current_best = -1;  /* -1 neither, 0 current best, 1 new */
+            /*
+             * If both have concrete and None in the same position and
+             * they are identical, we will continue searching using the
+             * first best for comparison, in an attempt to find a better
+             * one.
+             * In all cases, we give up resolution, since it would be
+             * necessary to compare to two "best" cases.
+             */
+            int unambiguous_equivally_good = 1;
+            for (Py_ssize_t i = 0; i < nargs; i++) {
+                int best;
+
+                /* Whether this (normally output) dtype was specified at all */
+                int is_not_specified = (
+                        op_dtypes[i] == (PyArray_DTypeMeta *)Py_None);
+
+                PyObject *prev_dtype = PyTuple_GET_ITEM(best_dtypes, i);
+                PyObject *new_dtype = PyTuple_GET_ITEM(curr_dtypes, i);
+
+                if (prev_dtype == new_dtype) {
+                    /* equivalent, so this entry does not matter */
+                    continue;
+                }
+                if (is_not_specified) {
```

Comment:
Yet this code seems covered, so I must be missing something... Or is it not actually covered?

Reply:
Yeah, it is covered. The new TODO, is a bit clearer maybe.

The uncovered parts have to do with abstract DTypes mainly (which is different from unspecified).  That means all the `abstract` and `subclass` paths are uncovered.